### PR TITLE
Reworked sending

### PR DESCRIPTION
**Pull request by JarvisCraft**
* Renamed to MessengerAdvancement as AdvancementAPI seems to be a better name for Singleton
+ Added .grant() and .revoke() methods
+ Added default values for variables
+ Added String-to-FrameType variants of input (supporting random values)
+ Returned setters without return type
+ Added @SuppressWarnings("unchecked") for .getJSON() method
- Deprecated .save() methods

### CODE REVIEW

File: AdvancementAPI.java [hunk 1]
```diff
@@ -71,66 +95,110 @@ public String getTitle() {
...
-    public AdvancementAPI withFrame(FrameType frame) {
+    public MessengerAdvancement withFrame(FrameType frame) {
         this.frame = frame;
         return this;
     }
 
+    public MessengerAdvancement withFrame(String frame) {
+        if (frame.equalsIgnoreCase("random")) {
+            FrameType[] frames = FrameType.values();
+            setFrame(frames[(int)(Math.random()*(frames.length-1))]);
```
Comment (yannicklamprecht) [inline AdvancementAPI.java]: move `frames[(int)(Math.random()*(frames.length-1))]` to FrameType. Then you're able to do FrameType.random()

File: AdvancementAPI.java [hunk 2]
```diff
@@ -71,66 +95,110 @@ public String getTitle() {
...
+    }
+
+    public void setFrame(FrameType frame) {
+        this.frame = frame;
+    }
+
+    public void setFrame(String frame) {
+        if (frame.equalsIgnoreCase("random")) {
+            FrameType[] frames = FrameType.values();
+            setFrame(frames[(int)(Math.random()*(frames.length-1))]);
```
Comment (yannicklamprecht) [inline AdvancementAPI.java]: same here -> setFrame(FrameType.random());

File: AdvancementAPI.java [hunk 3]
```diff
@@ -13,27 +15,37 @@
...
+
+    public MessengerAdvancement(NamespacedKey id) {
         this.id = id;
-        this.items = Lists.newArrayList();
-        this.announce = true;
-        this.toast = true;
+    }
+
+    public MessengerAdvancement(String id) {
+        this.id = new NamespacedKey(ContentMakersPlugin.getInstance(), id);
```
Comment (yannicklamprecht) [inline AdvancementAPI.java]: Won't work this way. You've to pass the the JavaPlugin. `ContentMakersPlugin.getInstance()` only works with your plugin.

File: AdvancementAPI.java [hunk 4]
```diff
@@ -44,25 +56,37 @@ public String getIcon() {
         return icon;
     }
 
-    public AdvancementAPI withIcon(String icon) {
+    public MessengerAdvancement withIcon(String icon) {
         this.icon = icon;
         return this;
     }
 
+    public void setIcon(String icon) {
```
Comment (yannicklamprecht) [inline AdvancementAPI.java]: Why adding this method? `withIcon(String icon)` does the work!

File: AdvancementAPI.java [hunk 5]
```diff
@@ -44,25 +56,37 @@ public String getIcon() {
...
         return description;
     }
 
-    public AdvancementAPI withDescription(String description) {
+    public MessengerAdvancement withDescription(String description) {
         this.description = description;
         return this;
     }
 
+    public void setDescription(String description) {
```
Comment (yannicklamprecht) [inline AdvancementAPI.java]: The same here. `withDescription(String description)` does it.

File: AdvancementAPI.java [hunk 6]
```diff
@@ -44,25 +56,37 @@ public String getIcon() {
...
+        this.description = description;
+    }
+
     public String getBackground() {
         return background;
     }
 
-    public AdvancementAPI withBackground(String url) {
+    public void setBackground(String background) {
```
Comment (yannicklamprecht) [inline AdvancementAPI.java]: Why renaming it? The name `withBackground` has a reason. It implies thats make use of the BuilderPattern.

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]
Needs some changes. An idea is to create a inner class for the Builder Pattern!

File: AdvancementAPI.java (hunk 3 above)
Comment (JarvisCraft) [reply AdvancementAPI.java]: Here my bad
Forgotten to change back to norm

Comment (JarvisCraft) [reply AdvancementAPI.java]: Will fix in a few minutes (when I'm from PC)

File: AdvancementAPI.java (hunk 4 above)
Comment (JarvisCraft) [reply AdvancementAPI.java]: About setters explained above - just to make the Api fit in cases when it's not necessary to have return
Though it's a thing to discuss

**Comment by JarvisCraft** [comment]
Will update to 1.2 soon fixing issues with wrong Plugin and Moving *random* to enum

**Comment by JarvisCraft** [comment]
As said before:
* replaced ContentMakers with JavaPlugin (plus changed this constructor)
+ made `RANDOM()` into enum inside of `FrameType`

Thanks to @ysl3000 

//PS setters are still a thing to discuss as `.with<NAME>()`'s don't seem to fit in those cases where value is only set for one param. I suggest:
making smth. like `.with<NAME>() { .set<NAME>(); return(this); }`
OR simply having only `.<NAME>()`'s
//PPS I'm also planning to add the `Map<String, AdvancementAPI>` (mentioned in Spigot discussion) as static member here and (probably) add "safe-show"

**Comment by yannicklamprecht** [comment]
Do you know Lombok ? I tested how good the Builder is working for our problem and it is working really well. https://github.com/Craftstuebchen/AdvancementAPI/blob/master/src/test/java/AdvancementAPITest.java

https://github.com/Craftstuebchen/AdvancementAPI/blob/master/src/main/java/io/chazza/advancementapi/AdvancementAPI.java

File: AdvancementAPI.java [hunk 7]
```diff
@@ -175,51 +242,112 @@ public String getJSON() {
...
+    @SuppressWarnings("deprecation")
+    public MessengerAdvancement remove() {
+        Bukkit.getUnsafe().removeAdvancement(id);
+        return this;
+    }
+
+    public MessengerAdvancement show(Player... players) {
+        add();
+        grant(players);
+        Bukkit.getScheduler().runTaskLater(ContentMakersPlugin.getInstance(), new Runnable() {
```
Comment (yannicklamprecht) [inline AdvancementAPI.java]: Hey I found the same problem here. `ContentMakersPlugin.getInstance()`

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]

File: AdvancementAPI.java (hunk 7 above)
Comment (JarvisCraft) [reply AdvancementAPI.java]: LoL
One minute

**Comment by JarvisCraft** [comment]
+ Fixed .send() 

Gonna work on shortening tomorrow

//Sorry, it's 1:58 in Moscow now

**Comment by JarvisCraft** [comment]
+ Added static `List<AdvancementAPI>` to hold all built Advancements generated during the following session
+ Added two static and one non-static `.build()` methods to replace all constructor-variants which also add AdvancementAPI-instances into static list
+ Added `.unbuild()` method to remove AdvancementAPI-instances from static list
+ Overwritten `.toString()` method`
+ Edited top JavaDocs (sorry if any mistakes)
* Renamed back into AdvancementAPI
* Renamed `.with<NAME>` methods into `.<NAME>` methods
- Removed AdvancementAPI(BukkitPlugin plugin, String name) constructor and made another one private
- Removed package
- Removed setters
- Removed unnecessary debug messages

**Comment by JarvisCraft** [comment]
The new way to create a custom Advancement to better fit the OneClass-BuilderPattern:
`//In the given example I am creating a new Advancement-Announcement which will have a Unique ID "<PLUGINNAME>:announcement_breaker", will be shown once (to players player1 and player2), will have the diamond axe as an icon and will have a bold dark_red description "BREAK'EM'ALL":
AdvancementAPI.build(plugin, "announcement_breaker")
                .title("§4§lBREAK'EM'ALL")
                .icon("minecraft:diamond_axe")
                .show(player1, player2)
                .unbuild();
//.unbuild() in the end also returns AdvancementAPI so you can use it again (though any changes in it require server restart/reload new Unique ID)`
For better code formatting see: https://pastebin.com/z1RrzNgL

**Pull request merged**